File: frontend/src/lib/conversations.ts
// Frontend API client for conversations
// lib/conversations.ts

const API_BASE = process.env.NEXT_PUBLIC_API_URL || 'http://localhost:8000';

// ═══════════════════════════════════════════════════════════════════
// TYPES
// ═══════════════════════════════════════════════════════════════════

export interface Conversation {
    id: string;
    title: string;
    created_at: string;
    updated_at: string;
}

export interface Message {
    id: string;
    conversation_id: string;
    role: 'user' | 'assistant';
    content: string;
    sources?: SourceItem[] | null;
    created_at: string;
}

export interface SourceItem {
    content: string;
    score: number;
    type: 'text' | 'image';
    metadata: Record<string, unknown>;
}

export interface ConversationWithMessages extends Conversation {
    messages: Message[];
}

// ═══════════════════════════════════════════════════════════════════
// API FUNCTIONS
// ═══════════════════════════════════════════════════════════════════

/**
 * Fetch all conversations (most recent first)
 */
export async function getConversations(limit = 50): Promise<Conversation[]> {
    const response = await fetch(`${API_BASE}/api/conversations?limit=${limit}`);

    if (!response.ok) {
        throw new Error('Failed to fetch conversations');
    }

    return response.json();
}

/**
 * Create a new conversation
 */
export async function createConversation(title?: string): Promise<Conversation> {
    const response = await fetch(`${API_BASE}/api/conversations`, {
        method: 'POST',
        headers: { 'Content-Type': 'application/json' },
        body: JSON.stringify({ title }),
    });

    if (!response.ok) {
        throw new Error('Failed to create conversation');
    }

    return response.json();
}

/**
 * Get a conversation with its messages
 */
export async function getConversation(id: string): Promise<ConversationWithMessages> {
    const response = await fetch(`${API_BASE}/api/conversations/${id}`);

    if (!response.ok) {
        if (response.status === 404) {
            throw new Error('Conversation not found');
        }
        throw new Error('Failed to fetch conversation');
    }

    return response.json();
}

/**
 * Add a message to a conversation
 */
export async function addMessage(
    conversationId: string,
    role: 'user' | 'assistant',
    content: string,
    sources?: SourceItem[]
): Promise<Message> {
    const response = await fetch(`${API_BASE}/api/conversations/${conversationId}/messages`, {
        method: 'POST',
        headers: { 'Content-Type': 'application/json' },
        body: JSON.stringify({ role, content, sources }),
    });

    if (!response.ok) {
        throw new Error('Failed to add message');
    }

    return response.json();
}

/**
 * Delete a conversation
 */
export async function deleteConversation(id: string): Promise<void> {
    const response = await fetch(`${API_BASE}/api/conversations/${id}`, {
        method: 'DELETE',
    });

    if (!response.ok) {
        throw new Error('Failed to delete conversation');
    }
}

/**
 * Update conversation title (auto-generate from first message)
 */
export function generateTitle(firstMessage: string): string {
    // Take first 50 chars or first sentence
    const truncated = firstMessage.slice(0, 50);
    const firstSentence = truncated.split(/[.!?]/)[0];
    return firstSentence.length > 0 ? firstSentence : truncated;
}

File: api/routes/conversations.py
# api/routes/conversations.py - REST API for conversation history
"""
Endpoints for managing conversations and messages.
Includes in-memory fallback when Supabase is not configured.
"""

from fastapi import APIRouter, HTTPException
from pydantic import BaseModel
from typing import List, Optional, Dict, Any
import json
import uuid
from datetime import datetime

router = APIRouter(prefix="/api/conversations", tags=["conversations"])


# ═══════════════════════════════════════════════════════════════════
# IN-MEMORY FALLBACK (when Supabase is not configured)
# ═══════════════════════════════════════════════════════════════════

_memory_conversations: Dict[str, Dict] = {}
_memory_messages: Dict[str, List[Dict]] = {}


def _is_supabase_available() -> bool:
    """Check if Supabase is configured and available."""
    try:
        from supabase_client import get_supabase
        client = get_supabase()
        return client is not None
    except:
        return False


# ═══════════════════════════════════════════════════════════════════
# MODELS
# ═══════════════════════════════════════════════════════════════════

class CreateConversationRequest(BaseModel):
    title: Optional[str] = None


class AddMessageRequest(BaseModel):
    role: str  # 'user' or 'assistant'
    content: str
    sources: Optional[List[Dict[str, Any]]] = None


class ConversationResponse(BaseModel):
    id: str
    title: str
    created_at: str
    updated_at: str


class MessageResponse(BaseModel):
    id: str
    conversation_id: str
    role: str
    content: str
    sources: Optional[List[Dict[str, Any]]] = None
    created_at: str


# ═══════════════════════════════════════════════════════════════════
# ENDPOINTS
# ═══════════════════════════════════════════════════════════════════

@router.get("")
async def list_conversations(limit: int = 50) -> List[Dict]:
    """Get all conversations, most recent first."""
    if _is_supabase_available():
        try:
            from supabase_client import get_conversations
            return await get_conversations(limit)
        except Exception as e:
            print(f"[WARN] Supabase error: {e}, using memory fallback")
    
    # In-memory fallback
    convs = list(_memory_conversations.values())
    convs.sort(key=lambda x: x.get("updated_at", ""), reverse=True)
    return convs[:limit]


@router.post("")
async def create_conversation(request: CreateConversationRequest) -> Dict:
    """Create a new conversation."""
    now = datetime.now().isoformat()
    
    if _is_supabase_available():
        try:
            from supabase_client import create_conversation as sb_create
            result = await sb_create(request.title)
            if result:
                return result
        except Exception as e:
            print(f"[WARN] Supabase error: {e}, using memory fallback")
    
    # In-memory fallback
    conv_id = str(uuid.uuid4())
    conv = {
        "id": conv_id,
        "title": request.title or f"Chat {datetime.now().strftime('%b %d, %H:%M')}",
        "created_at": now,
        "updated_at": now
    }
    _memory_conversations[conv_id] = conv
    _memory_messages[conv_id] = []
    return conv


@router.get("/{conversation_id}")
async def get_conversation(conversation_id: str) -> Dict:
    """Get a conversation with its messages."""
    if _is_supabase_available():
        try:
            from supabase_client import get_conversation_messages, get_supabase
            
            client = get_supabase()
            conv_result = client.table("conversations").select("*").eq("id", conversation_id).execute()
            
            if not conv_result.data:
                raise HTTPException(status_code=404, detail="Conversation not found")
            
            messages = await get_conversation_messages(conversation_id)
            
            return {
                **conv_result.data[0],
                "messages": messages
            }
        except HTTPException:
            raise
        except Exception as e:
            print(f"[WARN] Supabase error: {e}, using memory fallback")
    
    # In-memory fallback
    if conversation_id not in _memory_conversations:
        raise HTTPException(status_code=404, detail="Conversation not found")
    
    return {
        **_memory_conversations[conversation_id],
        "messages": _memory_messages.get(conversation_id, [])
    }


@router.post("/{conversation_id}/messages")
async def add_message(conversation_id: str, request: AddMessageRequest) -> Dict:
    """Add a message to a conversation."""
    now = datetime.now().isoformat()
    
    if _is_supabase_available():
        try:
            from supabase_client import add_message as sb_add, get_supabase
            result = await sb_add(
                conversation_id=conversation_id,
                role=request.role,
                content=request.content,
                sources=request.sources
            )
            if result:
                # Update conversation timestamp
                client = get_supabase()
                client.table("conversations").update({"updated_at": now}).eq("id", conversation_id).execute()
                return result
        except Exception as e:
            print(f"[WARN] Supabase error: {e}, using memory fallback")
    
    # In-memory fallback
    if conversation_id not in _memory_conversations:
        raise HTTPException(status_code=404, detail="Conversation not found")
    
    msg = {
        "id": str(uuid.uuid4()),
        "conversation_id": conversation_id,
        "role": request.role,
        "content": request.content,
        "sources": request.sources,
        "created_at": now
    }
    
    if conversation_id not in _memory_messages:
        _memory_messages[conversation_id] = []
    _memory_messages[conversation_id].append(msg)
    
    # Update conversation timestamp
    _memory_conversations[conversation_id]["updated_at"] = now
    
    return msg


@router.delete("/{conversation_id}")
async def delete_conversation(conversation_id: str) -> Dict:
    """Delete a conversation and all its messages."""
    if _is_supabase_available():
        try:
            from supabase_client import delete_conversation as sb_delete
            await sb_delete(conversation_id)
            return {"success": True}
        except Exception as e:
            print(f"[WARN] Supabase error: {e}, using memory fallback")
    
    # In-memory fallback
    if conversation_id in _memory_conversations:
        del _memory_conversations[conversation_id]
    if conversation_id in _memory_messages:
        del _memory_messages[conversation_id]
    
    return {"success": True}
